liquidhaskell-boot/src/Language/Haskell/Liquid/GHC/Plugin.hs

```diff
@@ -196,6 +199,40 @@ unoptimiseDynFlags df = updOptLevel 0 df
     , ghcMode      = CompManager
     }
 
+--------------------------------------------------------------------------------
+-- | Parsing phase -------------------------------------------------------------
+--------------------------------------------------------------------------------
+
+-- | We hook at this stage of the pipeline to store the parsed module
+-- for later processing during typechecking:
+--
+-- * Comments are preserved during parsing because we turn on
+--   'Opt_KeepRawTokenStream'. So we can get the /LH/ specs from there.
+--
+-- * The typechecker hook needs to do desugaring itself (see the
+--   comment on 'typecheckHook') which requires access to a full
+--   'TypecheckedModule'. Annoyingly, the typechecker hook is given
+--   only a 'TcGblEnv', so we have to redo typechecking there. And
+--   the input to that needs to be a 'ParsedModule'.
+--
+-- Since we have no good way of passing extra data from here to
+-- 'typecheckerHook', we resort to horrible global variables.
+--
+parsedHook :: [CommandLineOption] -> ModSummary -> ParsedResult -> Hsc ParsedResult
+parsedHook _ ms parsedResult = do
+  liftIO $ writeIORef parsedRef parsed
+  pure parsedResult
+  where
+    parsed = ParsedModule
+      { pm_mod_summary = ms
+      , pm_parsed_source = hpm_module . parsedResultModule $ parsedResult
+      , pm_extra_src_files = hpm_src_files . parsedResultModule $ parsedResult
+      }
+
+parsedRef :: IORef ParsedModule
+parsedRef = unsafePerformIO $ newIORef $ error "parsedResultAction hook didn't run"
```

Comment:
Is this global variable going to be a problem when multiple files are compiled concurrently?

Probably is should be a Set, and the plugin should remove modules from the set when they are no longer needed.

Reply:
OK I've added a new global mutable state that stores the LH breadcrumb for each module, shared with #2368.